[Documentation][feature/otx] Add Object Detection Tutorial (#1469)
open-edge-platform/training_extensions: docs/source/tutorials/base/how_to_train/detection.rst

Object Detection model
======================

This tutorial reveals end-to-end solution from installation to model export and optimization for object detection task on a certain example.
On this page we show how to train, validate, export and optimize ATSS model on WGISD public dataset.

> .. note::

Reviewer: How about making a bigger and bold font for this sentence?
![image](https://user-images.githubusercontent.com/80735344/213086206-75a9aa7e-a2e9-4f81-9925-b4feb4e396fd.png)

Author: It already is. In the next PR I'll leave an instruction, how to build this docs, my bad
![image](https://user-images.githubusercontent.com/33455576/213126691-719d7e2f-0c9d-4ebd-8705-2b0831a57c4a.png)

  To learn how to deploy the trained model, refer to: :doc:`../deploy`.

  To learn how to run the demo and visualize results, refer to: :doc:`../demo`.

The process has been tested on the following configuration.

- Ubuntu 20.04
- NVIDIA GeForce RTX 3090
- CUDA Toolkit 11.4 

*************************
Setup virtual environment
*************************

> You can follow the installation process from a :doc:`quick_start guide <../../../get_started/quick_start>` to create a universal virtual environment for all tasks. On the other hand, you can create a task-specific environment following the process below.

Reviewer: Is the below figure what we expected?
![image](https://user-images.githubusercontent.com/80735344/213085560-9103cc0c-6f09-498e-b742-41d06ff2a521.png)

Author: Yes, it's just GitHub can't render cross-pages references in RST properly :)

.. code-block::

    sudo apt-get install python3-venv
    # create virtual env
    bash ./otx/algorithms/detection/init_venv.sh det_venv
    # activate virtual env
    source det_venv/bin/activate

.. #TODO: Update installation part: virtual env, install prerequisite, pip install -e . or pip install otx

***************************
Dataset preparation
***************************

1. Clone a repository with `WGISD dataset <https://github.com/thsant/wgisd>`_. 

.. code-block::

  cd data
  git clone https://github.com/thsant/wgisd.git
  cd wgisd
  git checkout 6910edc5ae3aae8c20062941b1641821f0c30127

This dataset contains images of grapevines with the annotation for different varieties of grapes. 

- ``CDY``	- Chardonnay
- ``CFR``	- Cabernet Franc
- ``CSV``	- Cabernet Sauvignon
- ``SVB``	- Sauvignon Blanc
- ``SYH``	- Syrah

It's a great example to start with. The model achieves high accuracy righ from the beginning of the training due to relatively large and focused objects. Also, these objects are distinguished by a person, so we can check inference results just by looking at images.

> .. image:: ../../../../../utils/images/wgisd_dataset_sample.jpg

Reviewer: Need to be fixed
![image](https://user-images.githubusercontent.com/80735344/213085813-5dfec9ab-f4be-4df4-aace-58a1e7e80918.png)

Author: Sure

      └── test_bbox_instances.json
  ├──data/
      └── <images>
  ...

3. ``(Optional)`` To simplify the command line functions calling, we may create a ``data.yaml`` file with annotations info and pass it as a ``--data`` parameter. The content of the ``training_extesions/data.yaml`` for WGISD dataset should have absolute paths and will be similar to that:

.. code-block::

  {'data': 
    {'train': 
      {'ann-files': '/home/<username>/training_extensions/data/wgisd/coco_annotations/train_bbox_instances.json',
       'data-roots': '/home/<username>/training_extensions/data/wgisd/data'},
    'val':
      {'ann-files': '/home/<username>/training_extensions/data/wgisd/coco_annotations/test_bbox_instances.json',
       'data-roots': '/home/<username>/training_extensions/data/wgisd/data'},
    'test':
      {'ann-files': '/home/<username>/training_extensions/data/wgisd/coco_annotations/train_bbox_instances.json',
       'data-roots': '/home/<username>/training_extensions/data/wgisd/data'}
    }
  }

``Ann-files`` contains a path to the annotation, while ``data-roots`` is a path to the folder, where images are stored.

*********
Training
*********

> 1. First of all, we need to choose which object detection model will we train. The list of supported templates for object detection is available with the command line below. 

Reviewer: What do you think about the changing the indentation level?
![image](https://user-images.githubusercontent.com/80735344/213086002-8beba659-c5f6-4d44-a1c7-6ab0af312ee7.png)

Author: Good point with indents, but it's again the way how GitHub renders this. After building it looks much better :) 
![image](https://user-images.githubusercontent.com/33455576/213126338-8c8fc163-bf56-4c73-a342-16b853ba27f2.png)

Object Detection model
======================

> This tutorial reveals end-to-end solution from installation to model export and optimization for object detection task on a certain example.

Reviewer 2: ```suggestion
This tutorial reveals end-to-end solution from installation to model export and optimization for object detection task on a specific example.
```

On this page we show how to train, validate, export and optimize ATSS model on WGISD public dataset.

.. note::

  To learn how to deploy the trained model, refer to: :doc:`../deploy`.

  To learn how to run the demo and visualize results, refer to: :doc:`../demo`.

> The process has been tested on the following configuration.

Reviewer 3: Colon?

Author: Okay, but I'll fix it in the next documentation PR, if you don't mind

1. Check your python version:

.. code-block::

  python3 --version; pip3 --version; 

  Python 3.8.10

2. Create and activate a virtual environment for the obect detection task.

The following example creates a virtual environment in the ``det_venv`` folder for detection task.

.. code-block::

  sudo apt-get install python3-venv
  # create virtual env
  bash ./otx/algorithms/detection/init_venv.sh det_venv
  # activate virtual env
  source det_venv/bin/activate

.. #TODO: Update installation part: virtual env, install prerequisite, pip install -e . or pip install otx

***************************
Dataset preparation
***************************

1. Clone a repository with `WGISD dataset <https://github.com/thsant/wgisd>`_. 

.. code-block::

  cd data
  git clone https://github.com/thsant/wgisd.git
  cd wgisd
  git checkout 6910edc5ae3aae8c20062941b1641821f0c30127

This dataset contains images of grapevines with the annotation for different varieties of grapes. 

- ``CDY``	- Chardonnay
- ``CFR``	- Cabernet Franc
- ``CSV``	- Cabernet Sauvignon
- ``SVB``	- Sauvignon Blanc
- ``SYH``	- Syrah

> It's a great example to start with. The model achieves high accuracy righ from the beginning of the training due to relatively large and focused objects. Also, these objects are distinguished by a person, so we can check inference results just by looking at images.

Reviewer 2: ```suggestion
It's a great example to start with. The model achieves high accuracy right from the beginning of the training due to relatively large and focused objects. Also, these objects are distinguished by a person, so we can check inference results just by looking at images.
```

> .. image:: ../../../../utils/images/wgisd_dataset_sample.jpg

Reviewer 2: Can we add an image with bounding boxes? You wrote that "we can check inference results just by looking at images". I think after that it is a good idea to add image with inference results

Author: Done

.. code-block::

  {'data': 
    {'train': 
      {'ann-files': '/home/<username>/training_extensions/data/wgisd/coco_annotations/train_bbox_instances.json',
       'data-roots': '/home/<username>/training_extensions/data/wgisd/data'},
    'val':
      {'ann-files': '/home/<username>/training_extensions/data/wgisd/coco_annotations/test_bbox_instances.json',
       'data-roots': '/home/<username>/training_extensions/data/wgisd/data'},
    'test':
      {'ann-files': '/home/<username>/training_extensions/data/wgisd/coco_annotations/train_bbox_instances.json',
       'data-roots': '/home/<username>/training_extensions/data/wgisd/data'}
    }
  }

``Ann-files`` contains a path to the annotation, while ``data-roots`` is a path to the folder, where images are stored.

*********
Training
*********

1. First of all, we need to choose which object detection model will we train. The list of supported templates for object detection is available with the command line below. 

.. note::

>   The characteristics and detailed comparison of the models could be found in :doc:`Explanation section <../../../explanation/Main_algorithms/object_detection>`.

Reviewer 2: ```suggestion
  The characteristics and detailed comparison of the models could be found in :doc:`Explanation section <../../../explanation/algorithms/object_detection/object_detection.rst>`.
```

- ``weights.pth`` - a model snapshot
- ``label_schema.json`` - a label schema used in training, created from a dataset

These are needed as inputs for the further commands: ``export``, ``eval``,  ``optimize``,  ``deploy`` and ``demo``.

3. To have a specific example in this tutorial, all commands will be run on the ATSS model. It's a medium model, that achieves relatively high accuracy, while keeping the inference fast.

The following command line starts training of the medium object detection model on the first GPU on WGISD dataset:

.. code-block::

  (detection) ...$ otx train otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml
                            --train-ann-files data/wgisd/coco_annotations/train_bbox_instances.json 
                            --train-data-roots  data/wgisd/data
                            --val-ann-files data/wgisd/coco_annotations/test_bbox_instances.json 
                            --val-data-roots data/wgisd/data 
                            --save-model-to outputs
                            --work-dir outputs/logs
                            --gpus 1

> To start multi-gpu training, list the indexes of gpus you want to train on or omit `gpus` parameter, so training will run on all available GPUs.

Reviewer 2: ```suggestion
To start multi-gpu training, list the indexes of GPUs you want to train on or omit `gpus` parameter, so training will run on all available GPUs.
```

If you created ``data.yaml`` file in previous step, you can simplify the training by passing it in ``--data`` parameter:

.. code-block::

  (detection) ...$ otx train otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml
                            --data data.yaml
                            --save-model-to outputs
                            --work-dir outputs/logs
                            --gpus 1

Looks much simplier, isn't it?

> 4. ``(Optional)`` Additionally, we can tune training parameters such as batch size, learning rate, patience epochs or warm-up iteration. More about template-specific parameters is in quick start [#TODO link].

Reviewer 2: ```suggestion
4. ``(Optional)`` Additionally, we can tune training parameters such as batch size, learning rate, patience epochs or warm-up iterations. More about template-specific parameters is in quick start [#TODO link].
```

> It can be done by manually updating parameters in ``template.yaml`` file or via command line. 

Reviewer 2: ```suggestion
It can be done by manually updating parameters in the ``template.yaml`` file or via the command line.
```

> For example, to decrease batsch size to 4, fix the number of epochs to 100 and disable early stopping, extend the command line above with the following line.

Reviewer 2: ```suggestion
For example, to decrease the batch size to 4, fix the number of epochs to 100, and disable early stopping, extend the command line above with the following line.
```

.. code-block::

  (detection) ...$ otx eval otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml
                            --data data.yaml 
                            --load-weights outputs/weights.pth
                            --save-performance outputs/performance.json

We will get this validation output:

.. code-block::

  2023-01-10 06:17:32,182 | INFO : run task done.
  2023-01-10 06:17:33,884 | INFO : Inference completed
  2023-01-10 06:17:33,885 | INFO : called evaluate()
  2023-01-10 06:17:34,142 | INFO : F-measure after evaluation: 0.5487693710118504
  2023-01-10 06:17:34,142 | INFO : Evaluation completed
  Performance(score: 0.5487693710118504, dashboard: (1 metric groups))

3. The output of ``./outputs/performance.json`` consists of dict with target metric name and its value.

.. code-block::

  {"f-measure": 0.5487693710118504}

4. ``Optional`` Additionally, we can tune evaluation parameters such as confidence threshold via the command line. Read more about template-specific parameters for validation in quick start [#TODO link].

> For example, if there are too many False-Positive predictions (there we have prediction, but don't have annotated object for it) can suppress its number by increasing the confidence threshold as it is shown below.

Reviewer 2: ```suggestion
For example, if there are too many False-Positive predictions (there we have a prediction, but don't have annotated object for it) can suppress its number by increasing the confidence threshold as it is shown below.
```

> Please note, by default, the optimal confidence threshold is detected based on validation results to maximize the final F1 metric. So, to set a custom confidence threshold, please disable ``result_based_confidence_threshold`` option.

Reviewer 2: ```suggestion
Please note, by default, the optimal confidence threshold is detected based on validation results to maximize the final F1 metric. To set a custom confidence threshold, please disable the ``result_based_confidence_threshold`` option.
```

.. code-block::

  (detection) ...$ otx eval otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml
                            --data data.yaml 
                            --load-weights outputs/weights.pth
                            params 
                            --postprocessing.confidence_threshold 0.5
                            --postprocessing.result_based_confidence_threshold false 

  ...

  2023-01-10 06:21:04,254 | INFO : F-measure after evaluation: 0.514346439957492

*********
Export
*********

1. ``otx export`` exports a trained Pytorch `.pth` model to the OpenVINO™ Intermediate Representation (IR) format. 

> It allows to efficiently run it on Intel hardware, especially on CPU. Also, the resulting IR model is required to run POT optimization in the section below. IR model contains of 2 files: ``openvino.xml`` for weights and ``openvino.bin`` for architecture.

Reviewer 2: ```suggestion
It allows to run the model on the Intel hardware much more efficient, especially on the CPU. Also, the resulting IR model is required to run POT optimization in the section below. IR model consists of 2 files: ``openvino.xml`` for weights and ``openvino.bin`` for architecture.
```

> It allows to run the model on the Intel hardware much more efficient, especially on the CPU. Also, the resulting IR model is required to run POT optimization in the section below. IR model consists of 2 files: ``openvino.xml`` for weights and ``openvino.bin`` for architecture.

Reviewer 3: Much more efficient then what?

Reviewer 2: than Pytorch `.pth` model. It follows from the previous sentence.

Reviewer 3: The previous sentence just says that there are 2 model formats (although .pth is not really a format). Then, actually it's openvino runtime allows efficient inference of IRs, not the IR format itself. And I think we don't need an article before CPU, since we are not speaking about a particular one, but rather about Intel CPUs as a class of devices in general.

Author: I'll specify that OpenVINO runtime helps run the model faster, not the format itself,  but in my next documentation PR, if you don't mind. Thank you, this specification is really valuable

2. That's how we can export the trained model ``outputs/weights.pth`` from the previous section and save the exported model to the ``outputs/openvino/`` folder.

.. code-block::

  (detection) ...$ otx export otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml
                              --load-weights outputs/weights.pth
                              --save-model-to outputs/openvino/

  ...

  2023-01-10 06:23:41,621 | INFO : run task done.
  2023-01-10 06:23:41,630 | INFO : Exporting completed

> 3. We can check the accuracy of the IR model and the consistency between the exported model and the PyTorch model, using ``otx eval`` and passing IR model path to ``--load-weights`` parameter.

Reviewer 2: ```suggestion
3. We can check the accuracy of the IR model and the consistency between the exported model and the PyTorch model, using ``otx eval`` and passing the IR model path to the ``--load-weights`` parameter.
```

.. code-block::

  (detection) ...$ otx eval otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml
                            --test-ann-files data/wgisd/coco_annotations/test_bbox_instances.json 
                            --test-data-roots data/wgisd/data 
                            --load-weights outputs/openvino/openvino.xml
                            --save-performance outputs/openvino/performance.json
  
  ...
  2023-01-10 06:24:50,382 | INFO : Start OpenVINO inference
  2023-01-10 06:24:54,943 | INFO : OpenVINO inference completed
  2023-01-10 06:24:54,944 | INFO : Start OpenVINO metric evaluation
  2023-01-10 06:24:55,117 | INFO : OpenVINO metric evaluation completed
  Performance(score: 0.5487693710118504, dashboard: (1 metric groups))

*************
Optimization
*************

1. We can further optimize the model with ``otx optimize``. It uses NNCF or POT depending on the model format.

> ``NNCF`` optimization is used for trained snapshots in a framework-specific format such as checkpoint (pth) file from Pytorch. It starts accuracy-aware quantization based on the obtained weights from the training stage. Generally, we will see the same output as during training.

Reviewer 2: ```suggestion
``NNCF`` optimization is used for trained snapshots in a framework-specific format such as checkpoint (.pth) file from Pytorch. It starts accuracy-aware quantization based on the obtained weights from the training stage. Generally, we will see the same output as during training.
```

> ``POT`` optimization is used for models exported in the OpenVINO™ IR format. It decreases floating-point precision to integer precision of the exported model by performing the post-training optimization.

Reviewer 2: ```suggestion
``POT`` optimization is used for models exported in the OpenVINO™ IR format. It decreases the floating-point precision to integer precision of the exported model by performing the post-training optimization.
```

The function results with the following files, which could be used to run :doc:`otx demo <../demo>`:

- ``confidence_threshold``
- ``config.json``
- ``label_schema.json``
- ``openvino.bin``
- ``openvino.xml``

To learn more about optimization, refer to `NNCF repository <https://github.com/openvinotoolkit/nncf>`_.

> 2. Command example for optimizing a PyTorch model (`.pth`) with OpenVINO NNCF. We can also simplify the commanline by adding ``--data data.yaml`` parameter instead of specifying training and validation paths there.

Reviewer 2: ```suggestion
2. Command example for optimizing a PyTorch model (`.pth`) with OpenVINO™ NNCF. We can also simplify the command line by adding ``--data data.yaml`` parameter instead of specifying training and validation paths there.
```

.. code-block::

  (detection) ...$ otx optimize otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml 
                                --train-ann-files data/wgisd/coco_annotations/train_bbox_instances.json 
                                --train-data-roots  data/wgisd/data
                                --val-ann-files data/wgisd/coco_annotations/test_bbox_instances.json 
                                --val-data-roots data/wgisd/data 
                                --load-weights outputs/weights.pth
                                --save-model-to outputs/nncf
                                --save-performance outputs/nncf/performance.json

  ...

  2023-01-17 06:46:08,208 | INFO : run task done.
  2023-01-17 06:46:08,618 | INFO : Inference completed
  2023-01-17 06:46:08,618 | INFO : called evaluate()
  2023-01-17 06:46:08,829 | INFO : F-measure after evaluation: 0.5446735395189003
  2023-01-17 06:46:08,829 | INFO : Evaluation completed
  Performance(score: 0.5446735395189003, dashboard: (1 metric groups))

> 3. Command example for optimizing OpenVINO model (.xml) with OpenVINO POT. We can also simplify the commanline by adding ``--data data.yaml`` parameter instead of specifying training and validation paths there.

Reviewer 2: ```suggestion
3. Command example for optimizing OpenVINO™ model (.xml) with OpenVINO™ POT. We can also simplify the commanline by adding ``--data data.yaml`` parameter instead of specifying training and validation paths there.
```

> 3. Command example for optimizing OpenVINO model (.xml) with OpenVINO POT. We can also simplify the commanline by adding ``--data data.yaml`` parameter instead of specifying training and validation paths there.

Reviewer 2: BTW, we wrote about --data data.yaml 4 times above. I think it is redundant to add this info each time. What do you think of adding this + example for the train part and mentioning of course that we can use this for all otx command (eval, optimize)

Author: Done. But I still mention it in `otx eval` to highlight that it uses test paths, not val

  (detection) ...$ otx optimize otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml 
                                --train-ann-files data/wgisd/coco_annotations/train_bbox_instances.json 
                                --train-data-roots  data/wgisd/data
                                --val-ann-files data/wgisd/coco_annotations/test_bbox_instances.json 
                                --val-data-roots data/wgisd/data 
                                --load-weights outputs/openvino/openvino.xml
                                --save-model-to outputs/pot

  ...

  2023-01-10 06:29:46,751 | INFO : Loading OpenVINO OTXDetectionTask
  2023-01-10 06:29:47,685 | INFO : OpenVINO task initialization completed
  2023-01-10 06:29:47,685 | INFO : Start POT optimization
  2023-01-10 06:34:29,304 | INFO : POT optimization completed
  2023-01-10 06:34:29,419 | INFO : Start OpenVINO inference
  2023-01-10 06:34:33,275 | INFO : OpenVINO inference completed
  2023-01-10 06:34:33,275 | INFO : Start OpenVINO metric evaluation
  2023-01-10 06:34:33,451 | INFO : OpenVINO metric evaluation completed
  Performance(score: 0.5389435989256938, dashboard: (1 metric groups))

> The optimization time highly relies on the hardware characteristics, for example on 1 GeForce 3090 it took about 10 minutes.

Reviewer 3: depends on?

Author: It could be both

Please note, that POT will take some time without logging to optimize the model.

> 4. Finally, we can also evaluate the optimized model passing it to ``otx eval`` function.

Reviewer 2: ```suggestion
4. Finally, we can also evaluate the optimized model by passing it to the ``otx eval`` function.
```

> Now we have fully trained, optimized and exported in efficient model representation ready-to use object detection model.

Reviewer 2: ```suggestion
Now we have fully trained, optimized and exported an efficient model representation ready-to-use object detection model.
```

> Following tutorials provides further steps how to :doc:`deploy <../deploy>` and use your model in the :doc:`demonstration mode <../demo>` and visualize results.

Reviewer 2: ```suggestion
The following tutorials provide further steps on how to :doc:`deploy <../deploy>` and use your model in the :doc:`demonstration mode <../demo>` and visualize results
```
